063-topic-replication-factor-change.md

```diff
@@ -0,0 +1,221 @@
+
+The KafkaTopic resources marked as pending are added to an ongoingChanges map keyed by `User-Task-ID` when the topic configuration request is accepted by Cruise Control.
+This map is used to periodically check Cruise Control's tasks and update the KafkaTopic status.
+
+#### Configuration
+
+To request a replicas change, the Topic Operator needs to authenticate and send HTTP requests to the Cruise Control's REST API.
+The following environment variables are set automatically by the Cluster Operator when Cruise Control is configured in Kafka resource, or manually by the user in case of standalone deployment.
+
+| Name                                   | Type    | Default | Description                                                   |
+|:---------------------------------------|---------|---------|---------------------------------------------------------------|
+| STRIMZI_CRUISE_CONTROL_ENABLED         | Boolean | false   | Whether Cruise Control integration is enabled                 |
+| STRIMZI_CRUISE_CONTROL_NAMESPACE       | String  | ""      | Namespace where Cruise Control is running                     |
+| STRIMZI_CRUISE_CONTROL_HOSTNAME        | String  | ""      | Cruise Control hostname                                       |
+| STRIMZI_CRUISE_CONTROL_PORT            | Integer | 9090    | Cruise Control port                                           |
+| STRIMZI_CRUISE_CONTROL_SSL_ENABLED     | Boolean | false   | Whether SSL encryption is enabled                             |
+| STRIMZI_CRUISE_CONTROL_AUTH_ENABLED    | Boolean | false   | Whether Basic authentication is enabled                       |
+| STRIMZI_CRUISE_CONTROL_RACK_ENABLED    | Boolean | false   | Whether rack awareness is enabled in the target Kafka cluster |
+| STRIMZI_CRUISE_CONTROL_SECRET_NAME     | String  | ""      | Secret name containing the truststore                         |
+| STRIMZI_CRUISE_CONTROL_API_SECRET_NAME | String  | ""      | Secret name containing the API credentials                    |
+
+In case of Cluster Operator deployment, a dedicated admin user is automatically created for the Topic Operator.
+
+In case of standalone deployment, the Cluster Operator expects the following secrets to be created by the user:
+
+- A secret with `cruise-control.p12` key containing the truststore in PKCS12 format, and `cruise-control.password` key containing the truststore password.
+- A secret with `cruise-control.apiToAdminUsername` key containing the API username with admin role, and `cruise-control.apiToAdminPassword` key containing the user password.
+
+#### Error handling
+
+A replicas change is indefinitely retried by the periodic reconciliation when Cruise Control's task queue is full, or there is an error.
+The user is always informed about the replicas change state through appropriate logs and KafkaTopic status updates.
```

Comment:
So, how do you cancel or recover the change? Earlier, you suggested that once the RF is changed, any further changes are ignored. But if I by mistake change the the RF to 4 on my 3 node cluster, the change will be presumably failed by CC and I would want to cancel it (e.g. by changing the RF to some valid number). How would that work?

Reply:
All changes with errors, or completed with errors are automatically removed from pendingChanges and the user will be able to fix the error by updating the KT. If they don't do it, the TO will continue to retry (after some backoff period to not create too much work for CC) and log the same error.

Reply:
All changes with errors, or completed with errors are automatically removed from pendingChanges and the user will be able to fix the error by updating the KT. If they don't do it, the TO will continue to retry (after some backoff period to not create too much work for CC) and log the same error.